src/txrequest.h

```diff
@@ -0,0 +1,347 @@
+ *   ones), all of them are deleted (this is an invariant, and maintained by all operations below).
+ *
+ * The following operations are supported on this data structure:
+ *
+ * - ReceivedInv(txid, peer) adds a new CANDIDATE entry, unless one already exists for that (txid, peer) combination
+ *   (whether it's CANDIDATE, REQUESTED, or COMPLETED).
+ *
+ * - DeletedPeer(peer) deletes all entries for a given peer. It should be called when a peer goes offline.
+ *
+ * - AlreadyHaveTx(txid) deletes all entries for a given txid. It should be called when a transaction is successfully
+ *   added to the mempool, seen in a block, or for whatever reason we no longer care about it.
+ *
+ * - ReceivedResponse(peer, txid) converts any CANDIDATE or REQUESTED entry to a COMPLETED one. It should be called
+ *   whenever a transaction or NOTFOUND was received from a peer. When the transaction is acceptable, AlreadyHaveTx
+ *   should be called instead of (or in addition to) this operation.
+ *
+ * - GetRequestable(peer) does the following:
+ *   - Convert all REQUESTED entries (for all txids/peers) with (exptime <= now) to COMPLETED entries.
+ *   - Requestable txids are selected: txids which have no REQUESTED entry, and for which this peer has the
+ *     lowest-priority CANDIDATE entry among all CANDIDATEs with (reqtime <= now) at the time of the GetRequestable
+ *     call. As the high bit of the priority is whether or not the peer had a delay, immediate entries will always be
+ *     picked first as long as one is available.
+ *   - The selected txids are sorted in order of announcement and returned (even if multiple were added at the same
+ *     time, or even when the clock went backwards while they were being added).
+ *
+ * - RequestedTx(peer, txid) converts the CANDIDATE entry for the provided peer and txid into a REQUESTED one, with
+ *   exptime set to (now + timeout). It can ONLY be called immediately after GetRequestable was called (for the same
+ *   peer), with only AlreadyHaveTx and other RequestedTx calls (both for other txids) in between. Any other
+ *   non-const operation removes the ability to call RequestedTx.
+ */
+class TxRequestTracker {
+    //! Configuration option: timeout after requesting a download.
+    const std::chrono::microseconds m_timeout;
+
+    //! The various states a (txid,node) pair can be in.
+    //! Note that CANDIDATE is split up into 3 substates (DELAYED, BEST, READY), allowing more efficient implementation.
+    //! Also note that the sorting order of EntryTxid relies on the specific order of values in this enum.
+    enum class State : uint8_t {
+        //! A CANDIDATE entry whose reqtime is in the future.
+        CANDIDATE_DELAYED,
+        //! The best CANDIDATE for a given txid; only if there is no REQUESTED entry already for that txid.
+        //! The CANDIDATE_BEST is the lowest-priority entry among all CANDIDATE_READY (and _BEST) ones for that txid.
+        CANDIDATE_BEST,
```

Comment:
It's not that simple, unfortunately.

`GetRequestable` moves time (making things best/ready if they pass reqtime, making them completed if they pass exptime) for _all_ entries - not just those of the peer `GetRequestable` was called for. This is necessary, as otherwise you can't fairly decide what to give to which peer (the peer you're calling for may have an announcement that just became ready, but another peer may also have an announcement for the same tx with a lower priority). So it's conceptually true what you say, but only if you think of a full cycle of GetRequestable/RequestedTx for all peers.

I *really* want to avoid having TxRequestTracker depend on being able to call `AlreadyHave` - it would make it much less modularized and harder to test the data structure itself.